Add --classNames option to test command

janraasch opened the pull request:
This is continuing what we discussed in #40. I had some time on my hands during a longer train ride on the weekend. So I gave it a try.

# Feature

* `force-dev-tool test` accepts a `--classNames` option to execute specific test cases given by their names.
* `force-dev-tool test` accepts a `--verbose` option to show information about successful tests.

# Implementation

I rewrote `lib/cli/test` to handle both cases: *Run all tests* & *Run some tests*.

# Notes

## Modularization and function reuse

I rewrote `lib/cli/test` in a more functional manner (no need for inheritance here). The file now consists of a lot of small functions. Some of those could be extracted to `lib/utils` or something similar, e.g `printStartingMessage`, `logToConsole` and `login`. These functions could then be reused in other `lib/cli/*` files. 

## Testing

Also, other functions like `runTestsSynchronous` and `testClassNames` could be taken out to be tested in isolation. The great part about having functions like this, is that we can basically test a lot of the `cli` functionality without all the `Command`s and `Connection`s actually being set up. So we could e.g. mock a `connection` and then test `runTestsSynchronous`.

# Conclusion

I did not want to introduce to many new concepts here, I kept the code in `lib/cli/test` mostly untested except the `test-result-formatting` utility collection to stick with the existing style. I suggest to merge this and play with it to see, if you like the functional composition style of `lib/cli/test` and, if you feel like this could benefit the project, it would be an easy task to convert the other modules to rely on functional composition rather than protootypical inheritance.

# Sample Outputs

## `force-dev-tool test`

### Before

```bash
Running Test execution to remote dev

Error: Test execution failed.
 - System.NullPointerException: Attempt to de-reference a null object, stackTrace: Class.Test_QuoteController: line 76, column 1
 - System.NullPointerException: Attempt to de-reference a null object, stackTrace: Class.Test_QuoteController: line 76, column 1
 - System.NullPointerException: Attempt to de-reference a null object, stackTrace: Class.Test_QuoteController: line 76, column 1
 - Average test coverage across all Apex Classes and Triggers is 71%, at least 75% test coverage is required.
Visit https://SALESFORCEURL/changemgmt/monitorDeploymentsDetails.apexp?asyncId=SF_ID for more information.
```

### After

```bash
Running Test execution to remote dev

Failures:
Test_QuoteController#CreateAndSendQuoteTest took 1572.0
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Test_QuoteController#getCurrentUserName took 1338.0
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Test_QuoteController#getQuoteLineItemTest took 1315.0
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Code Coverage Warnings:
Average test coverage across all Apex Classes and Triggers is 71%, at least 75% test coverage is required.
Error: Visit https://cs83.salesforce.com/changemgmt/monitorDeploymentsDetails.apexp?asyncId=0Af4E000008Cv3SSAS for more information.
289 methods, 3 failures
```

## `force-dev-tool test --classNames 'Test_DecoratedOpportunity Test_QuoteController'`

```bash
Running Test execution to remote dev

Failures:
Test_QuoteController#CreateAndSendQuoteTest took 1536
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Test_QuoteController#getCurrentUserName took 1238
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Test_QuoteController#getQuoteLineItemTest took 1143
  - System.NullPointerException: Attempt to de-reference a null object
  - Class.Test_QuoteController: line 76, column 1
Error: 40 methods, 3 failures
```

Let me know, what you think.

Cheers!

amtrack commented:
Hi @janraasch ,
thank you so much! This looks like it is on a good way!

I will do a more detailed review but here is what i found by quickly trying this out.

When i run all tests for an org where there are no test errors, i get the following error:

```console
$ force-dev-tool test dev
Running Test execution to remote dev
/Users/me/src/github.com/amtrack/force-dev-tool/lib/test-result-formatting.js:34
  if (!failures.length) { return []; }
               ^

TypeError: Cannot read property 'length' of undefined
    at Object.formatFailures (/Users/me/src/github.com/amtrack/force-dev-tool/lib/test-result-formatting.js:34:16)
    at printTestAllResults (/Users/me/src/github.com/amtrack/force-dev-tool/lib/cli/test.js:98:40)
    at /Users/me/src/github.com/amtrack/force-dev-tool/lib/cli/test.js:121:3
    at /Users/me/src/github.com/amtrack/force-dev-tool/node_modules/jsforce/lib/promise.js:72:9
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

Could you please also run `npm run fmt` before committing? (I need to add this to a `CONTRIBUTING.md`)

janraasch commented:
Nice corner case :)

1. `Cannot read property 'length' of undefined``

I added some defensive null checks in https://github.com/amtrack/force-dev-tool/pull/42/commits/e504d7c323827276165a99a9edbdead388f292d9. We should put unit tests in place to lock this down. Can you provide a JSON `response` object for the request you did? (I do not have an org w/o tests handy to create and log a sample response myself.)

2. `npm run fmt`

I actually tried this, but it failed on me with

```bash
~/v/force-dev-tool> npm run fmt

> force-dev-tool@0.8.1 fmt /Users/Jan/force-dev-tool
> find lib test -name '*.js' | xargs --no-run-if-empty js-beautify -r

xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]

npm ERR! Darwin 15.6.0
npm ERR! argv "/Users/Jan/.nvm/versions/node/v6.2.1/bin/node" "/Users/Jan/.nvm/versions/node/v6.2.1/bin/npm" "run" "fmt"
npm ERR! node v6.2.1
npm ERR! npm  v3.9.3
npm ERR! code ELIFECYCLE
npm ERR! force-dev-tool@0.8.1 fmt: `find lib test -name '*.js' | xargs --no-run-if-empty js-beautify -r`
npm ERR! Exit status 1
```

Let me know, if you encounter more problems or have more feedback about the implementation details.

amtrack commented on lib/test-result-formatting.js:
```diff
@@ -0,0 +1,95 @@
+var formatCodeCoverageWarnings = function(testResult) {
+  var warnings = testResult.codeCoverageWarnings;
+
+  // TODO: What is this for? Or "When is this true?"
```
Yes, this is required! :-/
Unfortunately (all) the Metadata API responses return an object instead of an array if there is only one item. For the Tooling API this doesn't seem to be the case.

See https://gist.github.com/amtrack/d978987b1789e1f97c2285f6d29ecace for some sample responses for running `force-dev-tool test`.

So we should have a function s.th. like this

```javascript
var ensureArray = function(objectOrArray) {
  if (!objectOrArray) {
    return [];
  } else if (!Array.isArray(objectOrArray) {
    return [objectOrArray];
  }
  return objectOrArray;
};

var xxx = ensureArray(testResult.xxx);
if (!xxx.length) { return []; }
...
// return array with heading and lines
```

janraasch replied:
will add this to `utils`

amtrack commented on lib/test-result-formatting.js:
```diff
@@ -0,0 +1,95 @@
+    componentFailures = [componentFailures];
+  }
+
+  if (componentFailures.length) { return []; }
```
if (!componentFailures.length) { return []; }

janraasch replied:
yessir

amtrack commented on lib/test-result-formatting.js:
```diff
@@ -0,0 +1,95 @@
+};
+
+var formatOverview = function(testResult) {
+  var successCount = (testResult['successes'] && testResult['successes'].length) || 0;
```
mind that testResults['successes'] might be an object instead of an array.

So maybe s.th. like this:
```javascript
var successCount = ensureArray(testResult['successes']).length;
```

janraasch replied:
absolutely!

amtrack commented on lib/cli/test.js:
```diff
@@ -1,21 +1,191 @@
+
+		if (res.status === 'Failed') {
+			if (res.details && res.details.runTestResult) {
+				messages.push(formatting.formatOverview(res.details.runTestResult));
```
It would be good to always see the overview.
Move this before the `if (res.status === 'Failed') {`?

amtrack commented:
Hi @janraasch ,
i made my comments in line.

In total, i'm quite happy!
I played around with `--classNames` and `--verbose` and it is really useful! 👍 

FYI: I also merged #43 
So if you rebase your PR against master you should be able to run `npm run fmt` now.

For the backlog i'm thinking about treating the classnames as real argvs instead of a single arg.
This would allow things like `cat list_of_test_classes.txt | xargs force-dev-tool test --classNames`.
But to do that i will have to think about a concept for handling this in combination with the optional remote name...
Do you have any thoughts on that?
How would you do the call from your text editor (for multiple files)?

janraasch commented:
Hi @amtrack,

thank you for your detailed review. I agree with everything you said.

However, luckily I'll be on vacation for the next two weeks, so I won't be able to apply the changes you requested until after my vacation.

Just so you know :). If you'd like to do the changes yourself your welcome to do that, otherwise I'll try my best to work on this once I get back.

Kind regards,
Jan

amtrack commented:
There's no hurry.
I would prefer you finishing this when you come back, in order to get the credits. :-)
Enjoy your vacation!

janraasch commented:
As an aside:

*Südtirol kann ich nur empfehlen.*

Back to the issue at hand:

# your review remarks
- I agree with all of them
- changed the code **and** added more tests
- ran `npm run fmt`and it worked :)
- squashed everything into one commit and force pushed to this branch

# questions in your comment

> For the backlog i'm thinking about treating the classnames as real argvs instead of a single arg

Great idea.

> Do you have any thoughts on that?
> How would you do the call from your text editor (for multiple files)?

My use case mostly involves running the tests in **one** Test class, so this is an non-issue for me. thus, no thoughts at this point... :)

amtrack commented:
:tada: 
Merged and released in v0.10.0.

P.S.: Sorry for the delay! I have been on vacation as well.
Thank you very much for your great work!
